text/0000-dyn-error-generic-member-access.md

```diff
@@ -0,0 +1,471 @@
+- Feature Name: Add functions for generic member access to dyn Error and the `Error` trait
+- Start Date: 2020-04-01
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/2895)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes additions to the `Error` trait to support accessing generic
+forms of context from `dyn Error` trait objects. This generalizes the pattern
+used in `backtrace` and `source`. This proposal adds the method
+`Error::get_context` to the `Error` trait, which offers `TypeId`-based member
+lookup and a new inherent function `<dyn Error>::context` which makes use of an
+implementor's `get_context` to return a typed reference directly. These
+additions would primarily be useful for error reporting, where we typically no
+longer have type information and may be composing errors from many sources.
+
+_note_: This RFC focuses on the more complicated of it's two proposed
+solutions. The proposed solution provides support accessing dynamically sized
+types. The [alternative proposal] is easier to understand and may be more
+palatable.
+
+## TLDR
+
+Add this method to the `Error` trait
+
+```rust
+pub trait Error {
+    // ...
+
+    /// Provides an object of type `T` in response to this request.
+    fn get_context<'r, 'a>(&'a self, request: Request<'r, 'a>) -> ProvideResult<'r, 'a> {
```

Comment:
Ref https://github.com/rust-lang/rfcs/pull/2895#discussion_r403736887 @anp,

I find it strange that we have both `context()` function (a getter, sort of) and a `get_context()` function (sounds like a getter, but it's not). Keeping it `provide_context()` maybe better.